src/core/analytics.js

```diff
@@ -371,6 +377,98 @@ class Analytics {
     }
   }
 
+  /**
+   * A function to send single event to a destination
+   * @param {instance} destination
+   * @param {Object} rudderElement
+   * @param {String} methodName
+   */
+  sendDataToDestination(destination, rudderElement, methodName) {
+    try {
+      if (destination[methodName]) {
+        const clonedRudderElement = R.clone(rudderElement);
+        destination[methodName](clonedRudderElement);
+      }
+    } catch (err) {
+      const message = `[sendToNative]:: [Destination: ${destination.name}]:: `;
+      handleError(err, message);
+    }
+  }
+
+  /**
+   * A function to process and send events to device mode destinations
+   * @param {instance} destinations
+   * @param {Object} rudderElement
+   * @param {String} methodName
+   */
+  processAndSendEventsToDeviceMode(destinations, rudderElement, methodName) {
+    const destWithoutTransformation = [];
+    const destWithTransformation = [];
+
+    // Depending on transformation is connected or not
+    // create two sets of destinations
+    destinations.forEach((intg) => {
+      const sendEvent = !this.IsEventBlackListed(rudderElement.message.event, intg.name);
+
+      // Block the event if it is blacklisted for the device-mode destination
+      if (sendEvent) {
+        if (intg.areTransformationsConnected) {
+          destWithTransformation.push(intg);
+        } else {
+          destWithoutTransformation.push(intg);
+        }
+      }
+    });
+    // loop through destinations that doesn't have
+    // transformation connected with it and send events
+    destWithoutTransformation.forEach((intg) => {
+      this.sendDataToDestination(intg, rudderElement, methodName);
+    });
+
+    if (destWithTransformation.length > 0) {
```

Comment:
you can decouple to a sendTransformedDataToDestination method

Reply:
done